rust_src/src/dired.rs

```diff
@@ -0,0 +1,41 @@
+//! Lisp functions for making directory listings.
+
+use std::ffi::CStr;
+
+use libc;
+
+use remacs_macros::lisp_fn;
+use remacs_sys::Fuser_real_login_name;
+
+use lisp::{defsubr, LispObject};
+use lists::list;
+
+/// Return a list of user names currently registered in the system.
+/// If we don't know how to determine that on this platform, just
+/// return a list with one element, taken from `user-real-login-name'.
+#[lisp_fn]
+pub fn system_users() -> LispObject {
+    let mut done = false;
+    let mut unames = Vec::new();
+
+    while !done {
+        let pw: *mut libc::passwd = unsafe { libc::getpwent() };
+        if pw.is_null() {
+            done = true;
+        } else {
+            let pwnam_string =
+                unsafe { CStr::from_ptr((*pw).pw_name).to_string_lossy().into_owned() };
+            let pwnam_str = &*pwnam_string;
+            unames.push(LispObject::from(pwnam_str));
+        }
+    }
+    unsafe { libc::endpwent() };
+
+    if unames.len() < 1 {
+        unames.push(LispObject::from_raw(unsafe { Fuser_real_login_name() }));
```

Comment:
Is this equivalent? The C code is using a variable that the user can set, but this Rust version is calling a function.

Reply:
Yes, the function will return the current value of the variable.

I tried the variable first but it does seem to be visible (in src/editfns.o) to rust. nm(1) does not output it fwiw. So anyways it looked like I would have to make a change to editfns.c so was not worth it.

Reply:
In the C code a variable reference like `Vfoo` is really a macro that expands to `globals.f_Vfoo`, so in the Rust code we're currently just writing `globals.f_Vfoo` whenever we need to access the variable. The `globals` is still a global variable, but in the future it could be a thread-local variable or something.